src/sagemaker/image.py

```diff
@@ -0,0 +1,509 @@
+        """
+        if self.container:
+            self.container.down()
+        shutil.rmtree(self.container_root)
+
+    def retrieve_model_artifacts(self, compose_data):
+        """Get the model artifacts from all the container nodes.
+
+        Used after training completes to gather the data from all the individual containers. As the
+        official SageMaker Training Service, it will override duplicate files if multiple containers have
+        the same file names.
+
+        Args:
+            compose_data(dict): Docker-Compose configuration in dictionary format.
+
+        Returns: Local path to the collected model artifacts.
+
+        """
+        # Grab the model artifacts from all the Nodes.
+        s3_model_artifacts = os.path.join(self.container_root, 's3_model_artifacts')
+        os.mkdir(s3_model_artifacts)
+
+        for host in self.hosts:
+            volumes = compose_data['services'][str(host)]['volumes']
+
+            for volume in volumes:
+                container_dir, host_dir = volume.split(':')
+                if host_dir == '/opt/ml/model':
+                    self._recursive_copy(container_dir, s3_model_artifacts)
+
+        return s3_model_artifacts
+
+    def write_config_files(self, host, hyperparameters, input_data_config):
+        """Write the config files for the training containers.
+
+        This method writes the hyperparameters, resources and input data configuration files.
+
+        Args:
+            host (str): Host to write the configuration for
+            hyperparameters (dict): Hyperparameters for training.
+            input_data_config (dict): Training input channels to be used for training.
+
+        Returns: None
+
+        """
+
+        config_path = os.path.join(self.container_root, host, 'input', 'config')
+
+        resource_config = {
+            'current_host': host,
+            'hosts': self.hosts
+        }
+
+        json_input_data_config = {
+            c['ChannelName']: {'ContentType': 'application/octet-stream'} for c in input_data_config
+        }
+
+        _write_json_file(os.path.join(config_path, 'hyperparameters.json'), hyperparameters)
+        _write_json_file(os.path.join(config_path, 'resourceconfig.json'), resource_config)
+        _write_json_file(os.path.join(config_path, 'inputdataconfig.json'), json_input_data_config)
+
+    def _recursive_copy(self, src, dst):
+        for root, dirs, files in os.walk(src):
+            root = os.path.relpath(root, src)
+            current_path = os.path.join(src, root)
+            target_path = os.path.join(dst, root)
+
+            for file in files:
+                shutil.copy(os.path.join(current_path, file), os.path.join(target_path, file))
+            for dir in dirs:
+                new_dir = os.path.join(target_path, dir)
+                if not os.path.exists(new_dir):
+                    os.mkdir(os.path.join(target_path, dir))
+
+    def _download_folder(self, bucket_name, prefix, target):
```

Comment:
This is only weakly needed to be a method - consider making it a function that takes in a boto_session